Daily 리팩토링

theBettor opened the pull request:
## 🤷‍♂️ PR 내용(Issue 꼭 달기!)
resolved: #98 #106 #110 #111

1. 빠른 진행을 위해 뒤늦게 화면연결을 시도하였지만 서버 문제로 보여지지 않는 문제로 부분적으로 기능 구현함.
2. 당장 확인할 수 없는 부분을 제외하고는 디테일 잡아갈 예정

theBettor committed:
refactor : item들의 가로 꽉 채우기 #99

1. Column으로 감싸서 한 블록으로 만들기
2. start와 end 패딩을 통합하고, 너무 큰 end 패딩 (97.dp, 272.dp)을 제거 -> padding(horizontal = 20.dp) 적용

theBettor committed:
feat: daily-expand 모듈 추가 #103

theBettor committed:
fix: daily-manage 삭제 #103

하다보니까 preview 동작도 안하고해서 새로 만들기로함

theBettor committed:
refactor: daily가 아니라 dailyManage가 필요한것 같아 수정 #103

theBettor committed:
feat: daily-edit 모듈 추가 #105

근데 모듈 인식을 잘 못한다..

theBettor committed:
fix: 삭제된 buildgradle.kts 복구 #105

theBettor committed:
feat: edit 화면 dataPickerDialog 생성 #105

theBettor committed:
fix: datePick말고 TimePick #105

theBettor committed:
feat: dail-edit에 FilterChip 추가 #105

197~225번줄 코드를 추가하면 preview가 동작을 안합니다..!

theBettor committed:
feat: dail-edit에 FilterChip 추가

197~225번줄 코드를 추가하면 preview가 동작을 안합니다..!

theBettor committed:
Merge remote-tracking branch 'origin/refactor/daily' into refactor/daily

theBettor committed:
fix: 컴포넌트 구조 수정으로 preview 보이게하기 #109

IntrinsicSize.Min을 제거하고 fillMaxWidth와 height를 사용합니다.

theBettor committed:
fix: fiterChip 마무리

ChipDefaults.filterChipColors이 아니라 FilterChipDefaults를 써야함 왠만하면 material3쓰니가 둘의 차이를 명확히 구분해야함

theBettor committed:
fix: fiterChip 마무리

theBettor committed:
Merge remote-tracking branch 'origin/refactor/daily' into refactor/daily

theBettor committed:
refactor: DailyScreen 화면연결(with BottomNavigation) #107

theBettor committed:
refactor: DailyScreen 화면연결(with BottomNavigation)

theBettor committed:
Merge remote-tracking branch 'origin/refactor/daily' into refactor/daily

theBettor committed:
feat: 일과 수정 및 삭제할 수 있는 DropdownMenu 추가 #103

theBettor committed:
fix: 민준님의 DropdownMenu 따라해봄 #103

theBettor committed:
feat: DropdownMenu를 통한 일과 수정 및 삭제 시도 #107

theBettor committed:
feat: DropdownMenu를 통한 일과 수정 및 삭제 시도

theBettor committed:
Merge remote-tracking branch 'origin/refactor/daily' into refactor/daily

theBettor committed:
fix: lint

theBettor committed:
fix: lint

theBettor committed:
fix: 임시로 viewmodel 수정

DJDrama commented on feature/daily-edit/src/main/java/com/teampatch/feature/daily/edit/DailyEditSideEffect.kt:
```diff
@@ -0,0 +1,6 @@
+package com.teampatch.feature.daily.edit
+
+internal sealed interface DailyEditSideEffect {
```
DailyEditEvent 이름이 더 낳아보입니다.

왜나하면 사실상 Channel을 사용한다던지 해서 element를 내보내는 것이 SideEffect는 아니니깐요!

DJDrama commented on feature/daily-edit/src/main/java/com/teampatch/feature/daily/edit/DailyEditViewModel.kt:
```diff
@@ -0,0 +1,60 @@
+        private set
+
+    private val _sideEffect: Channel<DailyEditSideEffect> = Channel()
+    val sideEffect: Flow<DailyEditSideEffect> = _sideEffect.receiveAsFlow()
```
channel은 사실상 SideEffect가 아닌, state, event 를 관리하는 용도일테니, _event / event로 사용하면 좋지 않을까 싶습니다.

DJDrama commented on feature/daily-edit/src/main/java/com/teampatch/feature/daily/edit/DailyEditViewModel.kt:
```diff
@@ -0,0 +1,60 @@
+    val sideEffect: Flow<DailyEditSideEffect> = _sideEffect.receiveAsFlow()
+
+    var selectedDays = mutableStateOf(setOf<String>())
+        private set
```
위에 dailyEditUiState이 있다보니,
DailyEditUiState() data class 안에 selectedDays를 넣어서 처리하면 어떨까 싶습니다. 굳이 따로 필드를 하나 만들어서 할 필요 없이요!

DJDrama commented on feature/daily-expand/src/main/java/com/teampatch/feature/daily/expand/DailyExpandScreen.kt:
```diff
@@ -0,0 +1,251 @@
+    }
+
+    LaunchedEffect(Unit) {
+        viewModel.sideEffect.collect { sideEffect ->
```
viewModel.event.collect { ... }

로 하면 좋지 않을까 싶습니다~!

DJDrama commented on feature/daily-expand/src/main/java/com/teampatch/feature/daily/expand/DailyExpandScreen.kt:
```diff
@@ -0,0 +1,251 @@
+                verticalAlignment = Alignment.CenterVertically,
+                modifier = Modifier
+                    .fillMaxWidth()
+                    .padding(start = 24.dp, end = 24.dp, bottom = 16.dp)
```
.padding(horizontal = 24.dp)
.padding(bottom = 16.dp)

형태는 어떨까요?!

agvber commented on feature/daily-edit/src/main/java/com/teampatch/feature/daily/edit/DailyEditScreen.kt:
```diff
@@ -0,0 +1,289 @@
+    onConfirm: (hour: Int, minute: Int) -> Unit,
+) {
+    val context = LocalContext.current
+    val calendar = Calendar.getInstance()
```
Recomposition 될때 Instance 재생성 되는 성능 이슈가 있을 것 같아요!
remember 처리하시는 것은 어떨까요? ㄱㄱㄱ

agvber commented on feature/daily-expand/src/main/java/com/teampatch/feature/daily/expand/model/DailyExpandUiState.kt:
```diff
@@ -0,0 +1,11 @@
+import kotlinx.coroutines.flow.emptyFlow
+
+internal data class DailyExpandUiState(
+    val dailyManage: Flow<PagingData<DailyManage>> = emptyFlow(),
```
emptyFlow() 처리하면 crash 발생하지 않나요??

만약에 발생하면 flowOf(PagingData.empty()) 변경해주세요잇

agvber commented on feature/daily-expand/src/main/java/com/teampatch/feature/daily/expand/DailyExpandViewModel.kt:
```diff
@@ -0,0 +1,45 @@
+    private fun load() = viewModelScope.launch {
+        runCatching {
+            val daily = getDailyManageUseCase("someId") // 적절한 dailyId 값 사용
+            flowOf(PagingData.from(listOf(daily))) // DailyManage를 Flow<PagingData<DailyManage>>로 변환
```
잉 어짜피 List로 받으면 PagingData를 사용할 필요가 있을까요?
그리고 해당부분은 VIewModel에서 처리하기에는 문제가 있어보입니다!

agvber commented on feature/daily-expand/src/main/java/com/teampatch/feature/daily/expand/DailyExpandScreen.kt:
```diff
@@ -0,0 +1,251 @@
+                        .clickable { isDropDownMenuShow = true }
+                ) {
+                    Icon(
+                        painter = painterResource(com.teampatch.core.designsystem.R.drawable.ic_more_question),
```
import ㄱㄱ

agvber commented on feature/daily-edit/src/main/java/com/teampatch/feature/daily/edit/DailyEditScreen.kt:
```diff
@@ -0,0 +1,289 @@
+    }
+
+    LaunchedEffect(Unit) {
+        viewModel.sideEffect.collect {
```
https://github.com/Team-PATCH/Harmony-Android/pull/93#discussion_r1953794315

이 좋은 정보를 놓치다니 찬교님 후회 하실 뻔 했으요

agvber commented on feature/daily-edit/src/main/java/com/teampatch/feature/daily/edit/DailyEditScreen.kt:
```diff
@@ -0,0 +1,289 @@
+        viewModel.sideEffect.collect {
+            when (it) {
+                is DailyEditSideEffect.AddDailyError -> {
+                    Toast.makeText(context, "서버로 부터 데이터 전송 오류", Toast.LENGTH_SHORT).show()
```
저는 보통 에러 토스트 메시지는 LENGTH_LONG 처리 합니다!
왜냐하면 중요하다고 생각해서요
안드로이드 사용하시면 아실 것 같은데 TOAST 메시지가 은근 짧아요

agvber commented on feature/daily-edit/src/main/java/com/teampatch/feature/daily/edit/DailyEditScreen.kt:
```diff
@@ -0,0 +1,289 @@
+    )
+
+    LaunchedEffect(Unit) {
+        timePickerDialog.show()
```
음 해당 코드가 이해가 안되네요

시간 설정 버튼 클릭 -> TimePickerDialog가 보여야 될 것 같은데
코드가 위험 해보입니다!

@theBettor

agvber commented on feature/daily-edit/src/main/java/com/teampatch/feature/daily/edit/DailyEditScreen.kt:
```diff
@@ -0,0 +1,289 @@
+    onCompleteRequest: (String) -> Unit,
+    uiState: DailyEditUiState,
+) {
+    var daily by rememberSaveable { mutableStateOf(uiState.dailyExpand.content) }
```
어짜피 uiState에서 관리하고 있는데 굳이 또 선언 할 필요가 있을까요?

agvber replied:
Discord에서 구체적으로 말씀해달라고 해서 적습니다

## 문제

제가보기에는 2가지 문제가 보이는데

첫번째는 이미 uiState.dailyExpand.content는 이미 상태이므로 변경될때마다 Recomposition을 요청하므로 굳이 한번 더 mutableStateOf를 사용할 필요가 없다.

두번째는 rememberSaveable이 문제인데 Activity가 OnDestory -> OnCreate 되는 과정에서 savedInstanceState를 bunble 형태로 값을 저장 후 다시 restore되는 과정이 존재하는데
ViewModel에 존재하는 uiState 값은 위와 같은 생명주기가 진행되면 값이 초기화 되지만 rememberSaveable은 초기화 되지 않고 저장된 값을 불러오므로 uiState 와 daily가 값이 서로 일치하지 않을 가능성이 높습니다

![image](https://github.com/user-attachments/assets/41a39caa-0df3-49f7-b268-7339dad85741)


## 해결 

그래서 저라면 아래와 같이 ViewModel에서 상태 값을 변경할 것 같습니다.

```

class DailuViewModel {

   fun change블라블라(arg: String) {

     uiState.update { it.copy(블라블라 = arg) }  

   }
}



@Composable
internal fun DailyEditScreen(
    onDismissRequest: () -> Unit,
    onCompleteRequest: (String) -> Unit,
    onChange블라블라: (String) -> Unit,
    uiState: DailyEditUiState,
) {
```

agvber replied:
저도 틀릴 수 있으니 의심되면 튜터님에게 ㄱㄱㄱㄱㄱ

agvber commented on feature/daily-edit/src/main/java/com/teampatch/feature/daily/edit/DailyEditScreen.kt:
```diff
@@ -0,0 +1,289 @@
+            )
+
+            Row(
+                horizontalArrangement = Arrangement.SpaceEvenly, // 🔥 7개의 요일을 균등 배치
```
GPT 흔적쓰 ㅋㅋㅋ

agvber commented on feature/daily-edit/src/main/java/com/teampatch/feature/daily/edit/DailyEditScreen.kt:
```diff
@@ -0,0 +1,289 @@
+                verticalAlignment = Alignment.CenterVertically,
+                modifier = Modifier
+                    .fillMaxWidth()
+                    .height(52.dp) // IntrinsicSize 제거 후 명시적 높이 설정
```
아래 자식 레이아웃인 텍스트의 폰트 사이즈가 20.sp입니다.

혹시 폰트 크기를 최대로 하는 경우에 레이아웃이 문제 없는지 확인 해보셨나요?

agvber commented on feature/daily-edit/src/main/java/com/teampatch/feature/daily/edit/DailyEditViewModel.kt:
```diff
@@ -0,0 +1,60 @@
+import kotlinx.coroutines.launch
+
+/**
+ * DailyEdit를 위한 뷰모델과 Navigation이 덜 완성되었다... usecase도 더 만들어야될거같은데 민준님 도와주세요..
```
> usecase도 더 만들어야될거같은데 민준님 도와주세요..

어떤 부분이요 자세히 말해주세요!

agvber commented on feature/daily-edit/src/main/java/com/teampatch/feature/daily/edit/DailyEditViewModel.kt:
```diff
@@ -0,0 +1,60 @@
+        }
+    }
+
+    fun toggleDaySelection(day: String) {
```
이거 어디서 사용하는 건가요??

agvber commented on feature/daily-expand/src/main/java/com/teampatch/feature/daily/expand/DailyExpandScreen.kt:
```diff
@@ -0,0 +1,251 @@
+            )
+        }
+    ) { scaffoldPaddingValues ->
+        LazyColumn(
```
혹시 LazyColum에 Key를 넣을 수 있는 것을 아시나요??

https://youtu.be/EOQB8PTLkpY?t=251

agvber commented on feature/daily-edit/src/main/java/com/teampatch/feature/daily/edit/DailyEditScreen.kt:
```diff
@@ -0,0 +1,289 @@
+                    contentDescription = "time"
+                )
+                Text(
+                    text = time?.let { "${it.hour}:${it.minute}" } ?: stringResource(select_time),
```
이것도 가능하면 상태에서 처리 ㄱㄱㄱ

agvber commented on feature/daily-edit/src/main/java/com/teampatch/feature/daily/edit/DailyEditScreen.kt:
```diff
@@ -0,0 +1,289 @@
+            ) {
+                DayOfWeek.values().forEach { day ->
+                    FilterChip(
+                        selected = selectedDays.contains(day),
```
비용이 많이 들어가는 작업 처럼 보이는데 remember 처리하는 게 좋지 않을까요??

theBettor committed:
refactor: sideEffect가 아니라 Event로 이름 변경, selectDays 필드를 UiState에서 관리할 수 있게 변경

theBettor committed:
refactor: repeatOnLifecycle을 통하여 화면이 보이는 동안만 이벤트를 수집하여 성능 최적화

화면이 STOPPED 상태로 가면 이벤트 수집을 자동으로 중단하여 불필요한 리소스 사용을 방지함. 다시 STARTED로 돌아오면 이벤트 수집을 재개함.

theBettor committed:
refactor: recomposition시에도 인스턴스를 유지

remember를 통해 인스턴스를 한번만 생성하게 해서 재생성을 방지한다. Compose에서 특정 상태가 변경될 때, 해당 UI 부분만 다시 그리는 과정을 recomposition이라고 하는데 remember를 통해 불필요한 객체 재생성을 방지한다.

theBettor committed:
fix: 잘한건지는 모르겠는데 remember 처리 해주기

기존에 selectDays를 rememberSaveable로 유지하여 프로세스 정료 후에도 선택 상태를 복원할 수 있도록 하는데, 직접 참조하던 DayOfWeek.values()를 remember로 감싸 불필요한 values() 호출을 방지함으로써 리컴포지션을 최적화한다.

theBettor committed:
refactor: flowOf(PagingData.empty())로 변경 #115

theBettor committed:
Merge branch 'develop' into refactor/daily

theBettor committed:
Merge branch 'develop' into refactor/daily

theBettor committed:
fix: 뷰모델 임시 수정

theBettor committed:
Refactor: TimePickerDialog 수정

### 🔍 문제 분석 및 개선점
**1️⃣ TimePickerDialog가 즉시 실행되는 문제**
* LaunchedEffect(Unit) 내부에서 timePickerDialog.show()를 호출하면서, TimePickerDialog가 remember 시점에서 바로 실행됨.
* LaunchedEffect는 Composition이 시작될 때 실행되므로, TimePickerDialog가 조건 없이 실행되는 문제가 발생할 수 있음.
**🔹 해결책**
* LaunchedEffect를 제거하고, showDialog 상태를 if (showDialog) 블록으로 감싸 TimePickerDialog가 필요할 때만 실행됨.
- 불필요한 Composition 트리거 방지.

**2️⃣ Dialog가 닫힐 때** **onDismissRequest()****가 불필요하게 호출될 가능성**
* TimePickerDialog의 setOnDismissListener에서 onDismissRequest()를 호출하도록 설정했는데, 사용자가 onConfirm을 통해 시간을 선택할 때도 setOnDismissListener가 실행될 가능성이 있음.
* 결과적으로 onDismissRequest가 중복 호출되거나 의도치 않은 동작이 발생할 위험이 있음.

⠀**🔹 해결책**
* TimePickerDialog를 표시하는 로직을 isTimePickerDialogShow가 true일 때 실행되도록 제어하고, dismiss() 호출을 명확히 관리하고, onDismissRequest()가 setOnDismissListener 내부에서만 실행되도록 함

**3️⃣ 시간이 두 자리 수가 아닐 때 포맷 문제**
* 현재 time?.let { "${it.hour}:${it.minute}" }에서 hour와 minute 값이 한 자리 수일 경우 9:5와 같이 표시될 수 있음.
* 일반적인 UX 기대치는 09:05와 같이 두 자리 형식을 유지하는 것.

⠀**🔹 해결책**
* String.format("%02d:%02d", hour, minute)을 사용해 2자리 형식을 유지. 9:5가 아니라 09:05처럼 표시됨.

### 1.remember 최소화의 의미
remember를 최소화한다는 것은 **불필요한 상태 유지(state preservation)를 줄이는 것**을 의미해. remember는 Composable 함수가 재구성될 때 값을 유지하도록 도와주지만, 모든 상태를 remember로 관리할 필요는 없어.

✅ **이 코드에서 불필요한** **remember** **예시**
```kotlin
val calendar = remember { Calendar.getInstance() }
val hour = remember { calendar.get(Calendar.HOUR_OF_DAY) }
val minute = remember { calendar.get(Calendar.MINUTE) }
```

위 코드에서 calendar를 remember로 감싸는 것은 큰 의미가 없어. 왜냐하면 Calendar.getInstance()는 단순한 객체 생성을 수행할 뿐이고, 해당 값을 유지할 필요가 없기 때문이야. 대신 아래처럼 변경하는 게 좋겠지:

🔹 **개선된 코드**
```kotlin
val calendar = Calendar.getInstance()
val hour = calendar.get(Calendar.HOUR_OF_DAY)
val minute = calendar.get(Calendar.MINUTE)
```

이렇게 하면 calendar 자체를 remember할 필요 없이, 매번 새로운 값을 가져오게 돼.

**🚨 문제점:**
* TimePickerDialog는 **Android의** Dialog **객체이므로 리소스를 차지**하는데, 이 객체가 **재구성(recomposition)** 되어도 계속 살아남을 수 있어.
* TimePickerDialog가 **화면에서 사라질 때 명확하게 dismiss()를 호출하지 않으면** 메모리 누수(memory leak)나 예기치 않은 동작이 발생할 수 있어.

⠀✅ **해결책:** DisposableEffect **사용**
DisposableEffect를 사용하면 Composable이 **제거될 때(**onDispose**)** dismiss()**를 호출**할 수 있어.

🔹 **이렇게 개선하면?**
* show()를 별도로 호출하지 않아도 DisposableEffect 내부에서 자동으로 실행됨.
* Compose가 관리하는 리소스가 아니므로, **명확하게 해제하는 코드가 필요함** → onDispose 사용.
* DisposableEffect는 LaunchedEffect처럼 한 번 실행되지만, **recomposition이 발생해도 중복 실행되지 않음.** → 다이얼로그가 불필요하게 여러 개 생기지 않음.

### ✅ Recomposition이 되면 사라지는 것 vs. 사라지지 않는 것
맞아! Jetpack Compose에서 recomposition이 발생하면 **일부 값들은 새로 생성되고 사라지지만**, Android의 일부 객체들은 **계속 남아있을 수 있어.**
| **항목** | **Recomposition 시 유지됨?** | **설명** |
|:-:|:-:|:-:|
| val number = 10 | ❌ 사라짐 | 그냥 지역 변수이므로 recomposition 때마다 다시 초기화됨 |
| remember { mutableStateOf(0) } | ✅ 유지됨 | remember를 사용하면 상태를 유지할 수 있음 |
| DisposableEffect { MediaPlayer() } | ✅ 유지됨 (onDispose에서 해제해야 함) | MediaPlayer() 같은 리소스 객체는 **recomposition으로 사라지지 않음**, 직접 해제 필요 |
| TimePickerDialog(context, ...) | ✅ 유지됨 (onDispose에서 dismiss 필요) | Dialog는 Compose가 관리하는 게 아니므로 명시적으로 dismiss() 해줘야 함 |

### 📌 현재 코드 문제점
너의 기존 코드에서는 TimePickerDialog를 일반적인 Composable처럼 if (isTimePickerDialogShow) 안에서 직접 호출했어. 하지만 **다이얼로그는 상태 변화에 의해 재구성되면 안 돼!**
왜냐하면, Composable 함수는 Recomposition(재구성)이 일어날 수 있는데, 그럴 때마다 새로운 다이얼로그 객체가 계속 생성되면서 문제가 발생할 가능성이 높아.

# 🚀 1. Composable에서 제거한다는 의미
기존 코드처럼 TimePickerDialog를 Composable 안에서 호출하면 **재구성될 때마다 계속 생성**됨. 이를 방지하기 위해 Composable 함수에서 TimePickerDialog를 **없애고**, 일반 함수(function)로 만든다는 의미야.

### ✍️ 예제 (잘못된 코드)
```kotlin
if (isTimePickerDialogShow) {
    TimePickerDialog(
        onDismissRequest = { isTimePickerDialogShow = false },
        onConfirm = { hour, minute ->
            time = LocalTime.of(hour, minute)
            isTimePickerDialogShow = false
        }
    )
}
```

이렇게 Composable 안에서 다이얼로그를 직접 호출하면, isTimePickerDialogShow가 변경될 때마다 **계속 재구성됨**. 그래서 다이얼로그는 따로 일반 함수에서 관리해야 해!

# 🚀 2. DisposableEffect 기반으로 변경한다는 의미
### DisposableEffect는 **Side Effect(부수 효과)를 안전하게 관리**하는 Composable API야. **다이얼로그는 UI 요소가 아니라 부수 효과(Side Effect)이므로,** DisposableEffect**에서 다룬다**는 의미야.

## 🎯 결론
* ✅ **TimePickerDialogEffect****를 제거하고 일반 함수(****showTimePickerDialog()****)로 변경**
* ✅ **DisposableEffect** **안에서 다이얼로그를 관리 → UI 재구성과 무관하게 동작**
* ✅ **이제** **isTimePickerDialogShow****가** **true****일 때 한 번만 다이얼로그 실행됨**

theBettor committed:
fix : 안쓰는 뷰모델 함수 삭제

theBettor committed:
fix: uistate를 사용하고 있는것 같지 않아서 수정

현재 코드 흐름

1. DailyEditRoute에서 viewModel.dailyEditUiState를 uiState로 사용

2. DailyEditScreen에서 selectedDays를 uiState.selectedDays로 전달

3. onDaySelected로 ViewModel의 toggleSelectedDay(day)를 호출하여 상태 업데이트

agvber commented on feature/daily-edit/src/main/java/com/teampatch/feature/daily/edit/DailyEditScreen.kt:
```diff
@@ -0,0 +1,298 @@
+}
+
+@Composable
+fun TimePickerDialogEffect(
```
DisposableEffect 말고 그냥 아래와 같은 방식으로 처리하면 안되나요?
왜냐하면 isTimePickerDialogShow == true일때마다 새로운 인스턴스를 계속 참조하는게 좋지는 않아보입니다.

```
@Composable
internal fun DailyEditScreen(
) {

val timePickerDialog = remember {
TimePickerDialog(
            context,
            { _, selectedHour, selectedMinute -> onConfirm(selectedHour, selectedMinute) },
            hour,
            minute,
            true
        )
}


Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier
                    .fillMaxWidth()
                    .height(52.dp)
                    .background(color = WH, shape = RoundedCornerShape(10.dp))
                    .border(width = 1.dp, color = G2, shape = RoundedCornerShape(10.dp))
                    .padding(horizontal = 20.dp)
                    .noRippleClickable { timePickerDialog.show() }
            ) {
```

agvber commented on feature/daily-expand/src/main/java/com/teampatch/feature/daily/expand/DailyExpandScreen.kt:
```diff
@@ -0,0 +1,256 @@
+        )
+    }
+
+    LaunchedEffect(viewModel.event) {
```
이거 viewModel.event 값이 변경될때마다 LauchedEffect가 재실행 될 가능성이 높아보여요
어짜피 flow collect여서 Unit으로 처리해도 될 것 같습니다

agvber commented on feature/daily-expand/src/main/java/com/teampatch/feature/daily/expand/DailyExpandScreen.kt:
```diff
@@ -0,0 +1,256 @@
+
+@Composable
+fun DailyItem(
+    dailyItem: DailyManage?,
```
어짜피 Null checking을 앞에서 하는데 굳이 Null을 허용할 필요가 있을까요?

```
if (daily == null) {
                CircularProgressIndicator() // 로딩 상태 처리
            } else {
                DailyItem(
                    dailyItem = daily,
                    onEditDailyRequest = { onEditDailyRequest(daily) },
                    onDeleteDailyRequest = { onDeleteDailyRequest(daily) }
                )
            }
```

agvber commented:
수고하셨습니다! 
수정해야할 부분이 있을 것 같아서... 수정 후에 재요청해주세요!
(아니면 저의 틀린점을 반박해주시면 너무 좋습니다 ㅎㅎ)

theBettor committed:
refactor: DisposableEffect말고 remember 사용하기

remember를 활용해 TimePickerDialog 인스턴스를 재사용하면 DisposableEffect로 매번 새로 만드는 방식보다 메모리 측면에서도 깔끔하고 효율적입니다. 다만 remember는 Composable 내에서만 작동하고, context, hour, minute 등 초기화 시점에 필요한 값들이 유효한 시점에 있어야 해서 그 점도 고려해야 해요.TimePickerDialog를 remember로 컴포저블 안에서 관리하므로 DisposableEffect가 필요 없습니다.

theBettor committed:
refactor: flow collect여서 Unit으로 처리

1. viewModel.event가 바뀔 때마다 LaunchedEffect 블록이 재실행돼.
2. 굳이 viewModel.event를 key로 줄 필요 없고, 그냥 한 번만 실행되게 하면 됩니다~
3. Flow는 코틀린에서 비동기 데이터 스트림을 처리할 수 있는 도구야.
즉, 시간이 지남에 따라 데이터가 하나씩 emit(방출)될 수 있고, 우리는 그걸 collect() 해서 처리해.
viewModel.event.collect { event ->
    // 새로운 이벤트가 emit될 때마다 여기가 실행됨
}
→ 이건 "viewModel에서 이벤트가 나올 때마다 처리하자!"는 뜻이야.
4. Unit은 코틀린에서 "아무 값도 없어요!"를 나타내는 특별한 타입이야.
Unit을 LaunchedEffect의 key로 쓰면, 컴포저블이 처음 실행될 때만 딱 한 번 실행돼.

#요약
viewModel.event	: 변경될 수 있음 → 변경될 때마다 LaunchedEffect 재실행
Unit : 고정된 값 → LaunchedEffect는 한 번만 실행됨
collect : Flow에서 새 값이 emit될 때마다 자동 실행됨
5. 이렇게 하면 collect 안에서 알아서 이벤트를 처리하니까, 굳이 viewModel.event가 바뀔 때마다 전체 블록을 다시 실행할 필요가 없다는 뜻이야.

#LaunchedEffect
💡 LaunchedEffect는 언제 실행될까?
LaunchedEffect(key)는 이렇게 작동해:

컴포저블이 Composition에 처음 들어올 때 실행됨.

key가 변경될 때마다 다시 실행됨.

key가 같으면, Recomposition이 일어나도 재실행되지 않음.

🔁 예시로 살펴보자
kotlin
복사
편집
@Composable
fun MyScreen(userId: String) {
    LaunchedEffect(userId) {
        println("Fetching user info for $userId")
    }

    // 나머지 UI...
}
처음 MyScreen("123")이 그려질 때 → LaunchedEffect("123") 실행됨

userId가 "456"으로 바뀜 → LaunchedEffect("456") 실행됨 (다시!)

그냥 recomposition 발생 (예: 텍스트 입력 등, userId는 그대로) → ❌ 실행 안 됨

🔒 그래서 Unit을 key로 쓰면?
kotlin
복사
편집
LaunchedEffect(Unit) {
    // 처음 한 번만 실행
}
→ 이건 컴포저블이 처음 Composition에 들어올 때만 실행되고, 그 이후엔 절대 실행되지 않아
(심지어 recomposition이 수백 번 일어나도 다시 실행 안 됨)

✅ 이상적인 상황은?

상황	이상적인 key
단 한 번만 실행하고 싶다 (초기화, 초기 로딩 등)	Unit
특정 값이 바뀔 때마다 다시 실행하고 싶다	LaunchedEffect(key) 형태
Flow를 수집할 거다 (collect)	LaunchedEffect(Unit)로 한 번만 실행하고 collect {} 사용
💬 요약하면
LaunchedEffect는 "key 값"의 변화 기준으로 실행되며, Unit을 쓰면 처음 한 번만 실행되는 구조라 recomposition에도 영향을 받지 않아.
recomposition이 일어나더라도 key가 안 바뀌면 다시 실행되지 않음.

theBettor committed:
refactor: Null checking을 앞에서 하는데 굳이 Null을 허용할 필요가 없다.

theBettor committed:
fix: lint

theBettor committed:
refactor: "이미 상태인 값을 다시 상태로 만들면 상태 불일치가 생길 수 있으니, ViewModel의 상태만 사용하고, local state (예: rememberSaveable)는 피하자."

🔍 피드백 내용 요약
uiState.dailyExpand.content는 이미 상태이기 때문에
→ var daily by rememberSaveable { mutableStateOf(uiState.dailyExpand.content) }
❌ 이렇게 또 상태로 감싸지 말라는 의미야.

rememberSaveable은 생명주기 복원 시 ViewModel보다 우선순위가 높기 때문에
→ Activity가 재생성되면 ViewModel의 값은 새로 초기화되지만,
→ rememberSaveable은 복원된 값을 유지해서,
→ uiState.dailyExpand.content와 daily가 서로 다른 상태가 될 수 있다.

💡 요지는 이거야:
"이미 상태인 값을 다시 상태로 만들면 상태 불일치가 생길 수 있으니, ViewModel의 상태만 사용하고, local state (예: rememberSaveable)는 피하자."

✅ 결론: 너는 이렇게 수정하면 돼
rememberSaveable과 mutableStateOf를 daily에 사용하지 말고

uiState.dailyExpand.content만 읽고,

값 변경은 ViewModel 함수로 처리해.

3. val uiState by viewModel.dailyEditUiState에서 에러 발생 시 해결
이 부분은 State<DailyEditUiState>를 by로 위임하려면 dailyEditUiState가 State<DailyEditUiState> 타입이어야 해.

확인사항:

viewModel.dailyEditUiState가 val dailyEditUiState: State<DailyEditUiState> 혹은 val dailyEditUiState = mutableStateOf(...) 여야 by가 가능함.

예시로 ViewModel 코드:

kotlin
복사
편집
class DailyEditViewModel @Inject constructor(...) : ViewModel() {
    private val _dailyEditUiState = mutableStateOf(DailyEditUiState())
    val dailyEditUiState: State<DailyEditUiState> = _dailyEditUiState

    fun changeDailyContent(content: String) {
        _dailyEditUiState.value = _dailyEditUiState.value.copy(
            dailyExpand = _dailyEditUiState.value.dailyExpand.copy(content = content)
        )
    }

    // 이벤트, selectedDays 등 추가로 구현했겠지!
}

theBettor committed:
Merge branch 'develop' into refactor/daily
